crates/sui-network/src/discovery/mod.rs

```diff
@@ -0,0 +1,448 @@
+        );
+
+        for (peer_id, info) in eligible.into_iter().take(number_to_dial) {
+            let abort_handle = self
+                .tasks
+                .spawn(try_to_connect_to_peer(self.network.clone(), info));
+            self.pending_dials.insert(peer_id, abort_handle);
+        }
+
+        // If we aren't connected to anything and we aren't presently trying to connect to anyone
+        // we need to try the seed peers
+        if state.connected_peers.is_empty()
+            && self.pending_dials.is_empty()
+            && !self.config.seed_peers.is_empty()
+        {
+            let abort_handle = self.tasks.spawn(try_to_connect_to_seed_peers(
+                self.network.clone(),
+                self.config.seed_peers.clone(),
+            ));
+
+            // Use the all zero's peer_id as a marker for dialing seed peers
+            self.pending_dials.insert(PeerId([0; 32]), abort_handle);
+        }
+    }
+}
+
+async fn query_our_external_socket_address(peer: Peer, state: Arc<RwLock<State>>) {
+    let mut client = DiscoveryClient::new(peer);
+    let request = Request::new(()).with_timeout(TIMEOUT);
+    if let Ok(response) = client.get_external_address(request).await {
+        let addr = response.into_inner();
+        if let Some(our_info) = &mut state.write().unwrap().our_info {
+            our_info.external_socket_address = Some(addr);
+        }
+    }
+}
+
+async fn try_to_connect_to_peer(network: Network, info: NodeInfo) {
+    for multiaddr in &info.addresses {
+        if let Some(address) = multiaddr_to_anemo_address(multiaddr) {
+            // Ignore the result and just log the error if there is one
+            if network
+                .connect_with_peer_id(address, info.peer_id)
+                .await
+                .tap_err(|e| {
+                    debug!(
+                        "error dialing {} at address '{}': {e}",
+                        info.peer_id.short_display(4),
+                        multiaddr
+                    )
+                })
+                .is_ok()
+            {
+                return;
+            }
+        }
+    }
+
+    if let Some(socket_addr) = info.external_socket_address {
+        // Ignore the result and just log the error if there is one
+        let _ = network
+            .connect_with_peer_id(socket_addr, info.peer_id)
+            .await
+            .tap_err(|e| {
+                debug!(
+                    "error dialing {} at address '{}': {e}",
+                    info.peer_id.short_display(4),
+                    socket_addr
+                )
+            });
+    }
+}
+
+async fn try_to_connect_to_seed_peers(network: Network, seed_peers: Vec<SeedPeer>) {
+    let network = &network;
+
+    futures::stream::iter(seed_peers.into_iter().filter_map(|seed| {
+        multiaddr_to_anemo_address(&seed.address).map(|address| (seed, address))
+    }))
+    .for_each_concurrent(10, |(seed, address)| async move {
```

Comment:
why 10? seems low? maybe it doesn't matter much anyway :)

Reply:
I just picked a number, no real reason, although i should make it a constant so its clearer